File: backend/services/network_factory.py
from core.layers import DenseLayer
from core.network import MLP
from core.optimizers import get_optimizer

def build_network(architecture: list, weight_init: str):
    layers = []
    for prev_layer, layer in zip(architecture, architecture[1:]):
        layers.append(DenseLayer(
            input_size=prev_layer.units,
            output_size=layer.units,
            activation=layer.activation,
            weight_init=weight_init,
        ))
    return MLP(layers)

def build_optimizer(optimizer_name: str, lr: float):
    return get_optimizer(optimizer_name, lr)

File: netlab/src/engine/networkFactory.ts
import { DenseLayer } from './layers'
import { MLP } from './network'
import { getOptimizer, type OptimizerInstance } from './optimizers'
import type { LayerConfig, Optimizer, WeightInit } from '../types'

export function buildNetwork(architecture: LayerConfig[], weightInit: WeightInit): MLP {
    const layers: DenseLayer[] = []
    for (let i = 0; i < architecture.length - 1; i++) {
        const prevLayer = architecture[i]
        const layer = architecture[i + 1]
        if (!layer.activation) throw new Error(`Layer ${layer.id} is missing an activation`)
        layers.push(new DenseLayer(prevLayer.units, layer.units, layer.activation, weightInit))
    }
    return new MLP(layers)
}

export function buildOptimizer(name: Optimizer, lr: number): OptimizerInstance {
    return getOptimizer(name, lr)
}
